### ISSUE TITLE
Warn player they need to save money for VAB

### ISSUE THREAD

**Issue by peyre** [raceintospace/raceintospace#511]
Too many times I've forgotten my own advice and gone straight to R&D, then when I remember to go to the VAB to assemble a launch, I didn't save enough for the hardware.  I realize this wouldn't be trivial to implement, but it would be a nice feature if when you hit the RESEARCH PROGRAM FOR x MB button, the game would check to see if that would exceed the cost of assembling the missions you currently have in the VAB (as they are now, not taking into account the possibility of lowering costs by using, say, a B/Titan instead of a Saturn).

Or, perhaps, opening R&D/Purchasing would pop up a warning if you haven't assembled your hardware yet.

![NoMoney](https://user-images.githubusercontent.com/3118014/106363918-59940000-62e0-11eb-81d5-4aae0be0305f.png)

**Comment by peyre** [comment]
Got something!  I've committed a change that will bring up a warning if your cash is <= 60MB and you have at least one mission scheduled for this turn that hasn't been assembled yet.

![R DWarning](https://user-images.githubusercontent.com/3118014/164992944-23e816c6-c23c-4dae-9669-1b58e5487947.png)

**Comment by peyre** [comment]
Uh oh, I've missed something.  My launch on Pad A blew up the launch facility so it's canceled my Pad A mission for this turn--but R&D still thinks there's an unassembled mission on one of my pads.

![Destroyed Pad](https://user-images.githubusercontent.com/3118014/166156127-4b782664-254f-44a3-9004-68e4311e5c60.png)
![Warning Anyway](https://user-images.githubusercontent.com/3118014/166156130-9b07265c-fb3a-466d-808a-6ca53298d1e4.png)

**Comment by rnyoakum** [comment]
Can you save the game at this point and send me the file? Any mission data should have been cleared, so I would like to take a look at the file and see what's going on.

**Comment by peyre** [comment]
Oh!  Sure, will do.  I've just been looking at it; I think Data->P[player_index].LaunchFacility[i] might indicate a mission planned at low values, or cost to repair if high.

**Comment by peyre** [comment]
Here's the save file.

    On Sunday, May 1, 2022, 10:04:16 AM PDT, Ryan Yoakum ***@***.***> wrote:  
 
 


Can you save the game at this point and send me the file? Any mission data should have been cleared, so I would like to take a look at the file and see what's going on.

**Comment by rnyoakum** [comment]
Um, I don't see the save file attached - I'm thinking GitHub doesn't allow you to attach files of unknown types for security reasons. Can you email it to me directly?

**Comment by peyre** [comment]
Yes!  If I change it to Data->P[player_index].LaunchFacility[i] > 4 instead of > 0, it doesn't appear.

**Comment by rnyoakum** [comment]
> Oh! Sure, will do. I've just been looking at it; I think Data->P[player_index].LaunchFacility[i] might indicate a mission planned at low values, or cost to repair if high.

The `LaunchFacility` value is -1 if the launch pad is not built and 1 if it is built. If the pad is damaged or destroyed, the LaunchFacility value is set to the repair cost. So any value over 1 reflects damage to the pad.

**Comment by peyre** [comment]
Odd - the same variable means something else in another one of the files.  Having trouble finding it right now; I'm getting ready to reinstall my OS from scratch - I'm still on Xubuntu 20.10 and have been unable to upgrade for some time.  Now that the new LTS version (22.04) is out, I want to upgrade to that, but it won't go, so I have to wipe and reinstall from scratch.

**Comment by rnyoakum** [comment]
I opened up the save game file you sent me, and I'm not seeing an error. So, hopefully, this is fixed. Try it and see if everything works for you, like magic.

Nevertheless, I _think_ I know what happened with the save game, and if you pull up the save game you sent me, I don't think you'll see the warning anymore.

I'm guessing that last week when you were playing, you hadn't updated your build to include my commit of eleven days ago (d25229b95eae99d8fa5be6be71122175d2f9f90b). So in your implementation in 098752bed0eac822bdd97c0bb3678d7300d404af the code was

```
    // Warn player with little money that they should visit the VAB/VIB
    int Unass = 0;
    for (i = 0; i < 4; i++) {
        if (Data->P[player_index].LaunchFacility[i] > 0 && Data->P[player_index].Mission[i].MissionCode > 0 && Data->P[player_index].Mission[0].Hard[4] == 0) { Unass += 1; }
    }
    if (Unass > 0 && Data->P[player_index].Cash < 61) {
        Help("i173");
    }
```

which had a typo and **always** consulted the mission hardware for the mission on Pad A:
```
Data->P[player_index].Mission[0].Hard[4] == 0
```
Normally, if you had any missions scheduled for a turn there would be one on Pad A. But not that turn, because it exploded. So `Data->P[player_index].LaunchFacility[1] > 0` was **true** because you had a Pad B. Likewise,  `Data->P[player_index].Mission[i].MissionCode > 0` was **true** because a mission was scheduled and assembled on Pad B. Finally, `Data->P[player_index].Mission[0].Hard[4] == 0` was **true** because there was no Primary Booster assigned to the mission _on Pad A_.

But the newer commit fixes that, because I noticed the typo.

If that gets you down, I think I can make you feel better. Reviewing the code just now, I noticed I forgot a `break;` statement after calling the Help warning.

```
    // Warn player with little money that they should visit the VAB/VIB
    if (Data->P[plr].Cash <= 60) {
        for (int i = 0; i < MAX_MISSIONS; i++) {
            if (Data->P[plr].Mission[i].MissionCode > 0 &&
                Data->P[plr].Mission[i].Hard[Mission_PrimaryBooster] == 0) {
                Help("i173");
                // This should be a 'break;' command because the Help popup is only needed once.
            }
        }
    }
```

The player could conceivably receive the warning _three times_ if they have that many missions planned. So if you'd like, you can fix my oversight. Before I edit this post to forever hide my error.

<*Whistles...*>

**Comment by peyre** [comment]
I have my desktop mostly all back up and running now, so I can check this out now.  (rubs hands) 

I've been getting little smiles seeing that you've been at work on the game this weekend.  I do love when we make progress, even if it's only a little at a time.

So I've pulled up the save with the current version of the code, and you're right; I don't see the message when I open R&D, if the mission on Pad B is assembled.  If I unassemble it, R&D does pop up the warning, which is what I was hoping for.  So yes, I think we can consider this one taken care of--thanks for that!

Oh I see what you mean--you could get the warning once for every unassembled mission in the VAB.  I'll see if I can look into it, though today's sands are running out pretty quickly.

**Comment by peyre** [comment]
Ok, confirmed--you do see the warning once for every unassembled mission.  I'll see if I can fix it.

**Comment by peyre** [comment]
This warning thing seems to be working ok, but I think it would be better if it appeared only if the cost to purchase what's on the pad (with default rocket configuration) exceeds your cash by, say, 10 or 20 MB.

**Comment by peyre** [comment]
Here's a thought.  R&D and Purchasing could use all combined MIN COST figures of current missions (as discussed in #724) to decide whether to pop up this warning - rather than the crude <20 or <60 figure it uses at present.
